metadata-ingestion/src/datahub/ingestion/source/tableau.py

```diff
@@ -386,6 +454,125 @@ def _populate_usage_stat_registry(self):
+
+        def fetch_projects():
+            if self.server is None:
+                return all_project_map
+
+            for project in TSC.Pager(self.server.projects):
+                all_project_map[project.id] = TableauProject(
+                    id=project.id,
+                    name=project.name,
+                    parent_id=project.parent_id,
+                    description=project.description,
+                    path=[],
+                )
+
+        def set_project_path():
+            def form_path(project_id: str) -> List[str]:
+                cur_proj = all_project_map[project_id]
+                ancestors = [cur_proj.name]
+                while cur_proj.parent_id is not None:
+                    cur_proj = all_project_map[cur_proj.parent_id]
+                    ancestors.insert(0, cur_proj.name)
+                return ancestors
+
+            for project in all_project_map.values():
+                project.path = form_path(project_id=project.id)
+                logger.debug(f"Project {project.name}({project.id})")
+                logger.debug(f"Project path = {project.path}")
+
+        fetch_projects()
+        set_project_path()
+
+        return all_project_map
+
+    def _is_allowed_project(self, project: TableauProject) -> bool:
+        # Either project name or project path should exist in allow
+        is_allowed: bool = self.config.project_pattern.allowed(
+            project.name
+        ) or self.config.project_pattern.allowed(self._get_project_path(project))
+        if is_allowed is False:
+            logger.info(
+                f"project({project.name}) is not allowed as per project_pattern"
+            )
+        return is_allowed
+
+    def _is_denied_project(self, project: TableauProject) -> bool:
+        # Either project name or project path should exist in deny
+        for deny_pattern in self.config.project_pattern.deny:
+            # Either name or project path is denied
+            if re.match(
+                deny_pattern, project.name, self.config.project_pattern.regex_flags
+            ) or re.match(
+                deny_pattern,
+                self._get_project_path(project),
+                self.config.project_pattern.regex_flags,
+            ):
+                return True
+        logger.info(f"project({project.name}) is not denied as per project_pattern")
+        return False
+
+    def _populate_projects_registry(self):
+        if self.server is None:
+            return
+
+        logger.info("Populating site project registry")
+
+        all_project_map: Dict[str, TableauProject] = self._get_all_project()
+
+        def init_tableau_project_registry():
+            list_of_skip_projects: List[TableauProject] = []
+
+            for project in all_project_map.values():
+                # Skip project if it is not allowed
+                logger.debug(f"Evaluating project pattern for {project.name}")
+                if self._is_allowed_project(project) is False:
```

Comment:
Really? If allow pattern is not provided we should default to allow all?

Reply:
Will this do that?

Reply:
We always default to `default` project as projects configuration attribute is set to `default` project. some time later we do need to obsolete the `projects` key and need to request customer to update their recipe to use `project_pattern`